lib/puppet/application/resource.rb

```diff
@@ -142,10 +142,10 @@ def main
       resources = find_or_save_resources(type, name, params)
 
     if options[:to_yaml]
-      text = resources.map do |resource|
-        resource.prune_parameters(:parameters_to_include => @extra_params).to_hierayaml.force_encoding(Encoding.default_external)
```

Comment:
`force_encoding(Encoding.default_external)`

This might need to be added to your helper method or here in the new code. Encoding from external sources can sometimes be incorrectly set and then any string interpolation can be done incorrectly. We'll need to follow up on history of why we started using this initially and if it's still needed now.

Reply:
This code was added in 90891c9f72c441e0c8ee07347779baca6bc3c465 to fix a crash when concatenating strings with different encodings when building the YAML by hand.

Producing the document is now handled by the YAML library, and the unit test added to assert that mixing encodings does not raise exception in this commit is still passing with these changes.

So as far as I am concerned, we do not need extra care about the encoding at this level because the YAML library does the right thing.

Reply:
@smortex sadly there are 2 different issues going on. One is the commit you mentioned where `to_hierayaml` can return strings with different, but valid encodings, depending on how the provider retrieved the *current* value. For that issue, calling `force_encoding` isn't correct, if the `default_external` encoding doesn't match the string's encoding. I agree that letting `YAML.dump` handle that is correct:

```
irb(main):001:0> require 'yaml'
=> false
irb(main):002:0> utf8 = "José"
=> "José"
irb(main):003:0> latin1 = utf8.encode(Encoding::ISO_8859_1)
=> "Jos\xE9"
irb(main):004:0> YAML.dump([utf8, latin1])
=> "---\n- José\n- José\n"
```

However, there is a second issue that can occur if the provider returns a *current* property value whose encoding is incorrect. This used to happen with the user/group providers that use ruby's `Etc` module. We created a [`Puppet::Etc` wrapper](https://github.com/puppetlabs/puppet/blob/7944617322ea638cc546287c12a2be50d2e38d70/lib/puppet/etc.rb#L6-L17) and we  [scrub the resulting strings](https://github.com/puppetlabs/puppet/blob/7944617322ea638cc546287c12a2be50d2e38d70/lib/puppet/etc.rb#L166) so it is no longer an issue for those providers.

But it may happen for arbitrary providers, in which case `YAML.dump` will raise:

```
irb(main):023:0> utf8_jose = "José"
=> "José"
irb(main):024:0> jose_latin1 = utf8_jose.encode(Encoding::ISO_8859_1)
=> "Jos\xE9"
irb(main):025:0> jose_latin1_invalid = jose_latin1.dup.force_encoding('utf-8')
=> "Jos\xE9"
irb(main):026:0> jose_latin1_invalid.valid_encoding?
=> false
irb(main):027:0> YAML.dump(jose_latin1_invalid)
Traceback (most recent call last):
        6: from /usr/local/opt/rbenv/versions/2.5.3/bin/irb:11:in `<main>'
        5: from (irb):27
        4: from /usr/local/Cellar/rbenv/1.1.2/versions/2.5.3/lib/ruby/2.5.0/psych.rb:441:in `dump'
        3: from /usr/local/Cellar/rbenv/1.1.2/versions/2.5.3/lib/ruby/2.5.0/psych/visitors/yaml_tree.rb:118:in `push'
        2: from /usr/local/Cellar/rbenv/1.1.2/versions/2.5.3/lib/ruby/2.5.0/psych/visitors/yaml_tree.rb:136:in `accept'
        1: from /usr/local/Cellar/rbenv/1.1.2/versions/2.5.3/lib/ruby/2.5.0/psych/visitors/yaml_tree.rb:298:in `visit_String'
ArgumentError (invalid byte sequence in UTF-8)
irb(main):028:0>
```

To ensure we don't regress, I recommend having `to_hiera_yaml_hash` call `String#scrub` when serializing strings. That way it will replace invalid character sequences. Since we require ruby 2.3 or above in puppet#master, the `scrub` method should always be available.

Reply:
Please check c854b314fc5c954cc9e9230dbe1fc0445781e802 for a fix proposal.